codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -495,20 +518,22 @@ private void incrementPendingBytes(int numBytes) {
              * Note: this does not flush the {@link ChannelHandlerContext}.
              */
             int write(int allowedBytes) {
-                int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
                 try {
+                    pendingWriteQueue.remove();
```

Comment:
@nmittler - I agree that keeping the flow controller and encoder decoupled makes sense.  However it seems like there is an event that we are tip toeing around and adding work-arounds for.  When the write is done there is action that needs to be taken:

1. Removal from write queue. (done in flow controller)
2. If EOS notify life cycle manager to close local side. (done in encoder)

If this event needs to be distributed to different areas of the code the why not do that?  Set up an listener interface and have encoder implement it and the flow controller will notify this listener after it has done the queue cleanup.

> If the encoder closes/half-closes the stream (due to EOS or an error), this immediately triggers the flow controller to clear the queue for that stream (i.e. error any pending writes still in the queue). If the flow controller is in the middle of processing the pending write at the head of the queue when this happens, badness ensues (especially in the case of EOS, since we would write the stream and then kill it with RST_STREAM).

Unless we have successfully written the EOS and then proceed to write a RST I'm not sure I see a problem with this (which I think we may be able to avoid as described above?). What am I missing :)